### PR TITLE
Reduce unnecessary API calls

### PR DESCRIPTION
**Pull request by Integralist**
**Problem**: The top-level resources (e.g. fastly_service_vcl, fastly_service_compute, etc) only call the resource's associated 'Read' method (to update the internal Terraform state) if the resources were actually defined in the TF configuration, whereas nested block resources within a top-level resource (e.g. backend, domain, dictionary etc) have a `Read` method that will make GET API calls to update the data for that nested resource regardless of whether they were defined in the TF configuration.

**Solution**: Wrap each relevant section of code (in the nested block resource's 'Read' method) in a conditional statement that checks the length of available resources that exist within the state file.

Additionally, I discovered this implementation broke the provider's ability to import a service, and so I added a new computed attribute called `imported` that is temporarily set to `true` while an import is happening, and then reset it to `false` once the import is complete. This allows each nested resource's 'Read' method to not accidentally skip calling the API when an import is the intended behaviour.

**Results**:
This PR has caused the time it takes to run the end-to-end integration tests (as well as user `terraform plan`s) to be reduced by **~70%**!

**Notes**: There appears to be an inconsistency in how the resource 'key' is looked up. Most files use `GetKey()` while a couple of files use `Key()`. The former looks to be inherited by a base structure. I've not tried to clean this up, but I'm aware this might not be obvious to a code reviewer.

I've tested this code locally and can see that if I don't define a resource, then we don't try to refresh the state by calling the associated API endpoint. Then I define a new resource in the user Terraform configuration and I see that when running a `terraform plan` that the associated API endpoint is called for a refresh of state.

---

The review is easier with whitespace changes disabled...
https://github.com/fastly/terraform-provider-fastly/pull/593/files?w=1

### CODE REVIEW

**Comment by Integralist** [comment]
It would be good if someone (I'm thinking a seasoned user like @phamann) tried using this build with one of their projects to be sure nothing odd/incorrect happens.

**Comment by fgsch** [comment]
Might be easier to review adding `?w=1` to the url, e.g. `https://github.com/fastly/terraform-provider-fastly/pull/593/files?w=1`.

**Comment by Integralist** [comment]
The changes in the PR are effectively saying "when the 'Read' CRUD method is called, look at the state file, and see if there's anything in there yet - if not, skip the API call to refresh the state".

For a new project the 'Read' function is only called once there are resources created and put into the state file (e.g. the first `terraform plan` doesn't trigger 'Read', but after the initial resources are created all following `terraform plan`s will call 'Read' to update the state before making a comparison with the user's TF configuration).

If a user adds more resources to their TF configuration and run `terraform plan`, _then_ 'Read' is called and it will look at the state file and go (just as an example) "OK, you have no 'backends' in your state file so we're going to skip calling the Backends API looking for updated backends for this service".

That makes sense to me, but I guess the area I'm concerned with is where there are existing resources _before_ you set up your Terraform project. Would this PR cause a problem? I don't think it would because a user would have already run `terraform import` so the imported data would be placed into their state file.

But are there other instances where data would pre-exist and where `terraform import` wouldn't make sense? Is there a scenario I'm not thinking of where an API refresh would be needed and yet we skip it?

**Comment by jlav1n** [comment]
I tried it
```
│ Error: Request cancelled
│
│   with fastly_service_vcl.xxx,
│   on xxx.tf line 34, in resource "fastly_service_vcl" "xxx":
│   34: resource "fastly_service_vcl" "xxx" {
│
│ The plugin.(*GRPCProvider).ReadResource request was cancelled.
╵

Stack trace from the terraform-provider-fastly_v2.3.2-3-gb84a9730 plugin:

panic: interface conversion: interface {} is []interface {}, not *schema.Set

goroutine 15 [running]:
github.com/fastly/terraform-provider-fastly/fastly.(*WAFServiceAttributeHandler).Read(0xc0000ead00?, {0x1aae170?, 0xc000280ae0?}, 0xc0000ead00, 0xc0002cfb00, 0x0?)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/fastly/block_fastly_service_waf.go:103 +0x786
github.com/fastly/terraform-provider-fastly/fastly.resourceServiceRead({0x1aae170, 0xc000280ae0}, 0xc0000ead00, {0x180cc40?, 0xc000288180?}, {0x1aaad98, 0x1faeac0})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/fastly/base_fastly_service.go:516 +0xe42
github.com/fastly/terraform-provider-fastly/fastly.resourceRead.func1({0x1aae170?, 0xc000280ae0?}, 0x21bc108?, {0x180cc40?, 0xc000288180?})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/fastly/base_fastly_service.go:168 +0x35
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc00031b6c0, {0x1aae100, 0xc0002664c0}, 0x24?, {0x180cc40, 0xc000288180})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:358 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc00031b6c0, {0x1aae100, 0xc0002664c0}, 0xc00047c680, {0x180cc40, 0xc000288180})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:635 +0x35b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc0003128e8, {0x1aae100, 0xc0002664c0}, 0xc000266580)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:576 +0x43f
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc000311080, {0x1aae1a8?, 0xc00018a150?}, 0xc000278120)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server/server.go:553 +0x33c
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0x1933a40?, 0xc000311080}, {0x1aae1a8, 0xc00018a150}, 0xc000278000, 0x0)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:344 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000330a80, {0x1ab1858, 0xc00017c480}, 0xc000374400, 0xc0004d2690, 0x1fb68d0, 0x0)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:1194 +0xc8f
google.golang.org/grpc.(*Server).handleStream(0xc000330a80, {0x1ab1858, 0xc00017c480}, 0xc000374400, 0x0)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:1517 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:859 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:857 +0x28a

Error: The terraform-provider-fastly_v2.3.2-3-gb84a9730 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

**Comment by Integralist** [comment]
👋🏻 @jdigory thanks for taking a look. Interestingly I had noticed that myself when I was testing and thought I had removed the code from the WAF file(s), as it seemed to be using an unexpected data structure that didn't match all the other resources (and so I thought I would skip it until I figured out what its data structure was 😬) but seems your error is suggesting that `WAFServiceAttributeHandler` was the cause of a panic so maybe I left those changes in? 🤔

**Comment by Integralist** [comment]
@jdigory nope I totally left the code there 🤦🏻 (.e.g. the changes to `fastly/block_fastly_service_waf.go` should be removed from the change set for the moment).

**Comment by Integralist** [comment]
@jdigory just pushed commit to remove waf changes for the moment. let me know if the plan runs through for you now or if there are any other errors. Thanks!

**Comment by jlav1n** [comment]
> @jdigory just pushed commit to remove waf changes for the moment. let me know if the plan runs through for you now or if there are any other errors. Thanks!

Thanks! I pulled in your commit, `make build`, and tried `plan` again, but similar error
```
Stack trace from the terraform-provider-fastly_v2.3.2-3-gb84a9730 plugin:

panic: interface conversion: interface {} is []interface {}, not *schema.Set

goroutine 38 [running]:
github.com/fastly/terraform-provider-fastly/fastly.(*WAFServiceAttributeHandler).Read(0xc000480000?, {0x1aae170?, 0xc0001d8120?}, 0xc000480000, 0xc0008b8240, 0x0?)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/fastly/block_fastly_service_waf.go:103 +0x786
github.com/fastly/terraform-provider-fastly/fastly.resourceServiceRead({0x1aae170, 0xc0001d8120}, 0xc000480000, {0x180cc40?, 0xc0000af5c0?}, {0x1aaad98, 0x1faeac0})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/fastly/base_fastly_service.go:516 +0xe42
github.com/fastly/terraform-provider-fastly/fastly.resourceRead.func1({0x1aae170?, 0xc0001d8120?}, 0x21bc108?, {0x180cc40?, 0xc0000af5c0?})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/fastly/base_fastly_service.go:168 +0x35
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0002996c0, {0x1aae100, 0xc00069e500}, 0x24?, {0x180cc40, 0xc0000af5c0})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:358 +0x12e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc0002996c0, {0x1aae100, 0xc00069e500}, 0xc0001cc820, {0x180cc40, 0xc0000af5c0})
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource.go:635 +0x35b
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadResource(0xc000284900, {0x1aae100, 0xc00069e500}, 0xc00069e580)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/grpc_provider.go:576 +0x43f
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadResource(0xc0002c3000, {0x1aae1a8?, 0xc000452660?}, 0xc0000a66c0)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server/server.go:553 +0x33c
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadResource_Handler({0x1933a40?, 0xc0002c3000}, {0x1aae1a8, 0xc000452660}, 0xc0000a6660, 0x0)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:344 +0x170
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0002aea80, {0x1ab1858, 0xc000696000}, 0xc0000c6000, 0xc000452570, 0x1fb68d0, 0x0)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:1194 +0xc8f
google.golang.org/grpc.(*Server).handleStream(0xc0002aea80, {0x1ab1858, 0xc000696000}, 0xc0000c6000, 0x0)
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:1517 +0xa1b
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:859 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
        /Users/jlavin/go/src/github.com/fastly/terraform-provider-fastly/vendor/google.golang.org/grpc/server.go:857 +0x28a

Error: The terraform-provider-fastly_v2.3.2-3-gb84a9730 plugin crashed!
```
(Should it change the version string? It is the same as before...)

**Comment by jlav1n** [comment]
Ah, running `make clean` and then `make build` got it working! No more errors

**Comment by Integralist** [comment]
@fgsch after further testing + help testing from Josh I think this is ready for review.

**Comment by Integralist** [comment]
Also, I just tested a customer project (which was taking ~2m20s to run a plan) against this PR and I'm seeing the plan now takes 40s.

File: fastly/block_fastly_service_package.go [hunk 1]
```diff
@@ -74,24 +74,29 @@ func (h *PackageServiceAttributeHandler) Process(_ context.Context, d *schema.Re
...
-		return fmt.Errorf("error looking up Package for (%s), version (%v): %v", d.Id(), s.ActiveVersion.Number, err)
-	}
 
-	filename := d.Get("package.0.filename").(string)
-	wp := flattenPackage(pkg, filename)
-	if err := d.Set(h.GetKey(), wp); err != nil {
-		log.Printf("[WARN] Error setting Package for (%s): %s", d.Id(), err)
+		filename := d.Get("package.0.filename").(string)
+		wp := flattenPackage(pkg, filename)
+		log.Printf("[WARN] Package state: %+v", wp)
```
Comment (fgsch) [inline fastly/block_fastly_service_package.go]: Is this log call intended?

Comment (Integralist) [reply fastly/block_fastly_service_package.go]: Oh good spot! I'll remove that 🙂

Comment (Integralist) [inline fastly/block_fastly_service_package.go]:
```suggestion
```

**Review by fgsch — APPROVED** [review]
LGTM 👍

**Comment by Integralist** [comment]
There is a wrinkle in this change set: **importing a service no longer works**.

This is because a `terraform import` calls the 'Read' function for a resource so that it can populate the state, but our logic is saying "if there's nothing in the state then don't bother calling the API to get the latest data" (which is required for an import, i.e. although there is nothing in the state we _want_ 'Read' to go ahead in an import scenario).

I don't know at this point if there is a way for us to determine the 'Read' was called because of an import 😞

**Comment by Integralist** [comment]
OK fixed the import issue.

**Pull request merged**